### PR TITLE
[8.0][FIX] sale_delivery_split_date: sale_id not working

### PR DESCRIPTION
**Pull request by oihane**
When a picking is invoiced it will invoice using the delivery address (partner_id from stock.picking)
This fix avoids that error.

cc @tafaRU

### CODE REVIEW

**Comment by oihane** [comment]
@sbidoul @max3903 @pedrobaeza @mreficent @jbeficent 
Can anybody check this out? Thanks in advance.

File: sale_delivery_split_date/models/stock_picking.py [hunk 1]
```diff
@@ -3,16 +3,38 @@
...
+                line = line_obj.browse(cr, uid, line_id, context=context)
+                res[picking.id] = line.order_id
+        return res
+
     min_dt = fields.Date(
         string='Scheduled Date (for filter purpose only)',
         compute='_compute_min_dt', store=True)
 
+    _columns = {
+        'sale_id': old_fields.function(
```
Comment (MiquelRForgeFlow) [inline sale_delivery_split_date/models/stock_picking.py:33]: Why do you need to add this field? This field already exists.

Comment (oihane) [reply sale_delivery_split_date/models/stock_picking.py:33]: If I'm not mistaken when the field is a function (old API) you need to define again in order to extend the function itself. Not sure, and I haven't been able to prove.

Comment (MiquelRForgeFlow) [reply sale_delivery_split_date/models/stock_picking.py:33]: Ah, ok, I am not proficient with old api

**Review by MiquelRForgeFlow — APPROVED** [review]
LGTM

**Review by anajuaristi — APPROVED** [review]

**Comment by tafaRU** [comment]
@oihane, thanks for the PR. I'm hoping to review it in a day.

**Comment by tafaRU** [comment]
@oihane, could you please describe a bit better the use case this PR is intended to cover?
Thank you in advance.

**Comment by oihane** [comment]
I'll try

Usually, without this module, if you are using multiple addresses in sales orders you can define a delivery address and a different invoicing address
image

In that case picking will have Charlie Bernard and in invoice the address is going to be Jessica Dupont (when invoicing from picking or sale order)

Without this fix, if you invoice a picking the address of the invoice is Charlie Bernard and this PR fixes that.

Hope this helps a bit.

**Comment by tafaRU** [comment]
Thanks @oihane but I cannot understand what's the relation with this module :thinking: I mean, how is this relevant to splitting the delivery according to the requested date?
It would be better if you write the steps to reproduce it. Thanks!

**Comment by oihane** [comment]
I was wrong with the example.
First of all, with a demo order, that already has a picking, I've changed the payment term to "15 days" that is different from the one defined in the partner, by default it has "30 Net Days".
image

The invoice created from the transferred picking is:
image

Creating one with this module installed:
image
image

Then 2 delivery orders are created without sale.order
image

So when one of them is invoiced, the payment term is taken from the partner, not from the sale order
image

If you use the [sale_order_type](https://github.com/OCA/sale-workflow/tree/8.0/sale_order_type) module it will not pass order type from sale.order to stock.picking, and maybe there will be other modules also failing.

Maybe it is more clear now

File: sale_delivery_split_date/models/stock_picking.py [hunk 2]
```diff
@@ -3,16 +3,38 @@
...
+from openerp.osv import fields as old_fields
 
 
 class StockPicking(models.Model):
 
     _inherit = 'stock.picking'
 
+    def _get_sale_id(self, cr, uid, ids, name, args, context=None):
+        res = {}
+        line_obj = self.pool.get('sale.order.line')
```
Comment (tafaRU) [inline sale_delivery_split_date/models/stock_picking.py]: I'd prefer: `line_model = self.pool['sale.order.line']`

File: sale_delivery_split_date/tests/test_sale_delivery.py [hunk 3]
```diff
@@ -45,6 +45,10 @@ def test_check_single_date(self):
         self.assertEqual(
             len(self.so.picking_ids), 1,
             "There must be 1 picking for the SO when confirmed")
+        for picking in self.so.picking_ids:
+            self.assertEqual(
+                picking.sale_id, self.so,
+                "The sale order form picking must be the confirmed SO")
```
Comment (tafaRU) [inline sale_delivery_split_date/tests/test_sale_delivery.py]: could you rephrase it a bit? what's `form` stand for?

File: sale_delivery_split_date/tests/test_sale_delivery.py [hunk 4]
```diff
@@ -68,6 +72,10 @@ def test_check_multiple_dates(self):
         self.assertEqual(
             len(self.so.picking_ids), 2,
             "There must be 2 pickings for the SO when confirmed")
+        for picking in self.so.picking_ids:
+            self.assertEqual(
+                picking.sale_id, self.so,
+                "The sale order form picking must be the confirmed SO")
```
Comment (tafaRU) [inline sale_delivery_split_date/tests/test_sale_delivery.py]: same as above

File: sale_delivery_split_date/tests/test_sale_delivery.py [hunk 5]
```diff
@@ -101,6 +109,10 @@ def test_check_same_dates(self):
         self.assertEqual(
             len(self.so.picking_ids), 1,
             "There must be only one picking for the SO when confirmed")
+        for picking in self.so.picking_ids:
+            self.assertEqual(
+                picking.sale_id, self.so,
+                "The sale order form picking must be the confirmed SO")
```
Comment (tafaRU) [inline sale_delivery_split_date/tests/test_sale_delivery.py]: same as above

File: sale_delivery_split_date/models/stock_picking.py [hunk 6]
```diff
@@ -3,16 +3,38 @@
...
+        for picking in self.browse(cr, uid, ids, context=context):
+            res[picking.id] = super(StockPicking, self)._get_sale_id(
+                cr, uid, picking.ids, name, args, context=context)[picking.id]
+            if not res[picking.id] and picking.group_id:
+                line_id = line_obj.search(
```
Comment (tafaRU) [inline sale_delivery_split_date/models/stock_picking.py]: s/line_obj/line_model (see above)

**Review by tafaRU — CHANGES_REQUESTED** [review]
Thanks @oihane I finally got your point! 
Some comments, otherwise LGTM.

File: sale_delivery_split_date/tests/test_sale_delivery.py (hunk 3 above)
Comment (oihane) [reply sale_delivery_split_date/tests/test_sale_delivery.py]: misspelled, `from`

**Comment by oihane** [comment]
@tafaRU changes done

**Review by tafaRU — APPROVED** [review]
